c3d/src/C3d.cpp

```diff
@@ -432,16 +433,127 @@ class C3dParameterAttrInfo : public C3dSimpleSpelling, C3dDiagOnStruct {
 
         State = SeenExpr;
         return AttributeApplied;
+     }
+
+     default:
+        LLVM_DEBUG(llvm::dbgs() << "c3d: WARNING default case in handleDeclAttribute\n");
+        return NotHandled;
     }
   }
 };
 
+class C3dWhereAttrInfo : public C3dSimpleSpelling, C3dDiagOnStruct {
+public:
+  C3dWhereAttrInfo():
+    C3dSimpleSpelling("everparse::where"),
+    C3dDiagOnStruct("everparse::where")
+  {
+    NumArgs = 1;
+  }
+
+  AttrHandling parseAttributePayload(Parser *P,
+                                     ParsedAttributes &Attrs,
+                                     Declarator *D,
+                                     IdentifierInfo *AttrName,
+                                     SourceLocation AttrNameLoc,
+                                     SourceLocation *EndLoc,
+                                     IdentifierInfo *ScopeName,
+                                     SourceLocation ScopeLoc) const override {
+    // Note: we don't have access to the internal API of the Parser here, which
+    // makes things somewhat difficult. For instance, we don't have access to
+    // ConsumeParen(), or ExpectAndConsume()...
+    LLVM_DEBUG(llvm::dbgs() << "c3d: parsing argument to everparse::where\n");
+
+    // Defer to the parser to produce a good error message. TODO is this right?
+    if (P->getCurToken().getKind() != tok::l_paren)
+      return AttributeNotApplied;
+
+    // Eat opening left parenthesis.
+    P->ConsumeAnyToken();
+
+    ExprResult E;
+
+    // We push a new scope and introduce a dummy 'this' variable
+    // to parse the where clause. Then we pop it away, so it does
+    // not leak.
+    {
+        // Push a new scope. We use the DeclScope flag which permits
+        // declarations inside the new scope. Otherwise, we get
+        // an assertion failure at the ExitScope(). (An alternative
+        // is to manually delete the Decl from this new scope.)
+        P->EnterScope(Scope::DeclScope);
+
+        Sema &S = P->getActions();
+
+        // A 'void' type
+        const Type *voidTy = S.getASTContext().VoidTy.getTypePtr();
+        QualType R = QualType(voidTy, 0);
+
+        // A dummy 'this' identifier
+        IdentifierInfo &IDD = P->getPreprocessor().getIdentifierTable().getOwn("this");
+
+        // Create a variable declaration for 'this', at 'void' type, and push it.
```

Comment:
is there any way to factor out this bit into a helper function? (not sure) it appears in a few places in the code

Reply:
Yes, some bits of it at least, will do in future commits.